Potential 1.0.0

oriordan opened the pull request:
All right, this is a big one, apologies for that. Have been working on this for a few days and rebased it on your most recent changes. Happy to discuss anything, it's most likely breaking some of your existing stuff, but in general I think this brings the code to a shape which makes it much easier to use.

What I have not done but would seriously consider:
- Rename CodyAgent to AsyncCodyAgent
- Introduce a synchronous CodyAgent which uses AsyncCodyAgent under the hood

Most code is still sync out there and testing/prototyping async code is more pain. Since you went native async from day one, there is no need to make a sync copy but just wrapping around all methods in a sync equivalent works pretty well.

oriordan committed:
refactor(cody_py): Stripping down CodyServer class

- Simplified the case of testing node cody source code directly
- Removed the download functionality from CodyServer (will add to the
  cli script later on)
- Removed printing and added direct logging

feat(cli): Moved CLI script to the app

- Move the cli.py into the app and created script entry point in
  pyproject.toml. From now on you can `pip install .` and just use
  `codypy` command to run the CLI script.
- Added support for configuring custom enterprise Sourcegraph endpoint

refactor(logging): Using standard python logging
- Instead of the custom log function using standard python logging
- Configuring a logger on the CLI sciprt.

The changes aims to enhance usability of the module in other
applications and applying more common python patterns.

oriordan committed:
refactor(codypy): Restructure codypy module - breaking changes!

Removed files:
- Deleted config.py due to not being used anymore
- Deleted client_info.py and server_info.py and moved all their contents
  into models.py. This module contains all Pydantic BaseModels
- Deleted logger.py due to not being used anymore
- Deleted main.py. It's functionality has been somewhat re-written under
  cli.py
- Deleted context.py due to not being used anymore

Refactored & Created files:
- Refactored messaging.py: Created new RPCDriver class to wrap all the
  messaging functions to simplify it's use
- Created process.py: New class CodyProcessManager is responsible for
  the entire subprocess management
- Refactored server.py: CodyServer is now responsible for handling the
  CodyProcessManager object and the RPCDriver object. It is also alone
  responsible for initializing the Agent and to Shutdown the connection
- Refactored agent.py: Since CodyAgent is what users are interfacing
  with most of the time, init() call moved to this object from the
  CodyServer. This somewhat simplifies the user of the library. It also
  supports instantiating the Agents with some sane defaults instead of
  forcing users to create an AgentSpecs manually
- Created chat.py: A new Chat class is now returned when
  agent.new_chat() is being called. This is because one agent can have
  multiple chat session open the same time, and each has it's own context.
  Some methods which are context aware have moved to Chat class like
  set_model() and set_repository_context().
  Chat.ask() method is returning a Transcript object which is just a
  Pydantic model around the response dict, which contains the messages,
  context files, and a timestamp. This is to be improved later on, but
  the goal was to return "raw" messages.
- Refactored cli.py: This file is moved to the project and when it's
  pip-installed, the `codypy` command will act as running cli.py. It
  does support 3 operations:
  1. ask: Ask a single question
  2. chat: Start a chat session. Similar to what main() did before.
  3. download: It performs the download of cody-agent binary

Additional changes:
- Adding pylint into pre-commit
- Adding pylint for Github actions

This refactor aims to provide a more pythonic interface and (I think)
somewhat simplifies the use of the library. The goal was to let users
the minium amount of requires parameters while instantiating a new Agent
and handle multiple Chat sessions the same time.

One extra feature has been added to restore chats from a previous
conversation. This requires to pass the previous messages to
re-establish the context.

PriNova commented:
Hey @oriordan, thank you for your contributions.

This is a significant change, and as I haven't written a contribution guide, I'll leave it open for code review and add comments sporadically when time permits.
As I'm currently addressing some issues in the code, anticipate potential breaking changes and rebase as necessary.

In the future, I would prefer PRs that focus on smaller, single-issue changes.
I would also appreciate an issue request being opened beforehand to discuss any enhancements or additions to the codebase.

Please don't be discouraged if you wish to contribute further. For substantial changes, it's advisable to open a draft PR and incrementally discuss the changes.

Thank you for your time. I really appreciate it.

oriordan committed:
feat(models): Modelling contextFile objects according to cody-agent specs

Instead of letting Message.contextFile being list[Any] we're now
modelling all the different context item types according to
https://github.com/sourcegraph/cody/blob/main/lib/shared/src/codebase-context/messages.ts

PriNova commented:
To focus on small PR changes and on specific context, may you open a new PR where you can integrate the logging behavior only into the codebase? Is this OK for you?

oriordan commented:
Sure, can do that tonight. So just to agree:
- Dropping `debug_method_map` completely: I don't think stable code needs this
- Dropping is_debugging flag: this should be controlled by standard logging loglevels and not extra flags
- Dropping logger.py and all of it's references

If you want to keep something similar to `debug_method_map`, I suggest to create a new logger for every method and send logs via that. This allows to have a very granular logging control over the different methods. Example:
```python
method_logger = logging.getLogger(f"{__name__}.{method}")
method_logger.debug("Request: %s %s", method, params)
method_logger.debug("Response: %s", response)
```

Then if someone wants to only debug certain methods, they can just enable/disable the required loggers only.

wdyt?

PriNova commented:
The `debug_method_map` was implemented to disable stream messages when sending a chat request. I think in the future we may implement a stream chat function to handle this correctly.
Let us have the logger as a Singleton, where the user can toggle the debugging. For now, the logging in the `chat` function need to be disabled for the `webpost_request` (sorry, forget the name, but can be inspected in the function).
To implement the logger functionality, we can put it in a `logger.py` file.

I appreciate your work.

oriordan committed:
feat(messaging): Using configurable read-timeout & enhance logging

Using configurable read timeout (defaults to 10s, same as the cody
agent) instead of hard-coded 5s due to intermittently hitting timeout
for no apparent reason against enterprise server. Also adding some
logging when we hit timeout including the last message we received.

oriordan committed:
fix(process): Passing all environment variables

The original code updated the envvars during execution. This was later
changed to only pass the debug flag and the tcp server flag, however
this breaks the node process as libuv requires `HOME` envvar to be
present (and who knows what else). For doing the right thing here and
fot future proofing this, we now copy all envvars and add the ones which
we explicitly control via application parameters.

oriordan committed:
feat(messaging): Fix non-ASCII character handling in issue #10

Previously, the length of messages was calculated based on the number of
Unicode characters, which caused issues when sending non-ASCII
characters to Cody. This was due to the discrepancy between the length
of a Unicode string and the length of the same string represented as
bytes.

This commit fixes the issue by calculating the message length as bytes,
which accurately represents the length regardless of the characters
used. This change prevents exceptions in the Cody agent binary when
non-ASCII characters are included in messages.

oriordan committed:
feat(agent): Sanitize messages for chat restore

When restory a chat session, Cody expects a list of dictionaries which
contains two keys: "text" and "speaker". Previously we passed Message
objects which also implements these fields however it also contained
contextFiles. If we just pass these, the contextField field will eat up
tokens and trigger context limit violations. Also this field is
technically irrelevant for the conversation thread so we're ensuring
it's not sent while restoring a session.

oriordan committed:
feat(messaging): Using native json.loads to deserialise response

Previously we used pydantic_core.from_json which can't decode
non-complete unicode characters:

>>> pydantic_core.from_json('{"foo": "\\ud83c"}')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: unexpected end of hex escape at line 1 column 16

Using native json module is slightly slower but it'll be able to
deserialise partial unicode responses.

oriordan committed:
fix(messaging): Ensure only one coroutine can read/write socket

Since this is asynchronous code, in some truly async application use it
happened that multiple coroutines tried to read/write the socket the
same time. This is of course not possible if the socket hasn't finished
sending data. For making this concurrency safe, we're using a lock to
ensure only a single coroutine can read/write at a given time, and block
the others until it's done.
